README.md
[![Build Status](https://travis-ci.org/stojanovic/vreme.svg)](https://travis-ci.org/stojanovic/vreme) [![npm version](https://badge.fury.io/js/vreme.svg)](http://badge.fury.io/js/vreme)

<h1 align="center">
  <img width="320" src="https://rawgit.com/stojanovic/vreme/master/vreme.svg" alt="vreme">
  <br>
</h1>

Format date and time by providing human-readable example.

Vreme is a simple date/time converter with human readable format strings.

This module is mostly for simple and readable formatting, you should consider 
some other module for more complex date formats (ie. [this one](https://www.npmjs.com/package/strftime))

"Vreme" means Time in Serbian.

## Usage

Install module via:

`npm install vreme`

or

`bower install vreme`.

Then import Vreme and make a new instance, than you should simply provide date and 
human readable format.

If works with node.js and in the browser. It supports CommonJS, AMD and using as
global variable.

You can format full dates and times or just a month/day name, as you can see 
in the example below:

```

var Vreme = require('Vreme')

var vreme = new Vreme()

var date = new Date('Sat, 01 Aug 2015 14:10:21')

// To get month
console.log(vreme.format(date, 'January'))            // Output 'August'
console.log(vreme.format(date, 'Jan'))                // Aug
console.log(vreme.format(date, '12'))                 // 08
console.log(vreme.format(date, '2'))                  // 8

// To convert day names
console.log(vreme.format(date, 'Monday'))             // Saturday
console.log(vreme.format(date, 'monday'))             // saturday
console.log(vreme.format(date, 'Mon'))                // Sat
console.log(vreme.format(date, 'Mo'))                 // Sa
console.log(vreme.format(date, 'MO'))                 // SA
console.log(vreme.format(date, '1st'))                // 1st
console.log(vreme.format(date, '22'))                 // 01

// To get year
console.log(vreme.format(date, '1999'))               // 2015
console.log(vreme.format(date, '42'))                 // 15

// To get time
console.log(vreme.format(date, '4:10 pm'))            // '2:10 pm'
console.log(vreme.format(date, '15:32'))              // '14:10'

// To get full dates
console.log(vreme.format(date, 'March 25, 1999'))     // August 01, 2015
console.log(vreme.format(date, 'March 1, 1999'))      // August 1, 2015
console.log(vreme.format(date, 'March 25th'))         // August 1st
console.log(vreme.format(date, '2014/04/25'))         // 2015/08/01
console.log(vreme.format(date, '2014/04/25 4:10 am')) // 2015/08/01 2:10 pm
console.log(vreme.format(date, '02/03/11'))           // 08/01/15
console.log(vreme.format(date, '21.04.2015'))         // 01.08.2015

// Or you can combine it with text
console.log(vreme.format(date, 'Date: March, 25th'))  // Date: August, 1st


```

### With prototype

If you pass `usePrototype: true` in options Vreme will be added as a method
of Date object (`Date#formatLike`), so you can use it like this:

```

var Vreme = require('Vreme')

var vreme = new Vreme({
  usePrototype: true
})

var date = new Date('8/20/2015')

// Then you can get date like this
console.log(date.formatLike('March 25, 1999'))    // August 20, 2015


```

### In the browser

Just include vreme.js or vreme.min.js in scripts and use the module same as 
described above.

ie.

```

<!doctype html>
<html>
<head>
  <title>Vreme test</title>
</head>
<body>
  <script src="vreme.min.js"></script>
  <script>

    var vreme = new Vreme();

    var date = new Date();

    console.log(vreme.format(date, 'Monday'));

  </script>
</body>
</html>


```

## Limitations

This module is relying on regular expressions and I tried to keep it as simple 
as it's possible, so there is some limitations:

- If you are using integers smaller than 13 for month, year and day (ie. 
'02/03/11') module will always assume `MM/DD/YY` format;
- If you use integers between 13 and 31 for both year and day it'll assume (ie.
'02/14/14') it'll assume DD/YY format;
- At the moment it works only with month and day names in english.

There are additional limitation that are not on this list, they'll be added.

## Development

Vreme is written in ES6 and compiled with Babel.  
Intall Babel by running `npm i babel -g` than run `npm run compile` to 
transpile it to ES5.  

If you want to run Babel manually, you'll need to do: 
`babel src/index.js -o .tmp/vreme.js` and then `buildify` to 
add support for CommonJS, AMD and script and to make minified version.

## Test

Run `npm test` or `mocha -R spec ./test/index.js`.

## Todo

- [x] Parse time too;
- [x] Make it work in the browser without browserify;
- [ ] Memoization;
- [ ] Support other languages, it'll be a bit problematic to fix ordinal sufixes for
all languages, but everything else should be fine.


CHANGELOG.md
__Vreme__

#   (2016-01-08)



---

## Bug Fixes

- **babel:** migrate to babel 6 and es6 import
  ([ca763ebf](https://github.com/stojanovic/vreme/commit/ca763ebf4fea35bc7de95c52983cf5af63fdfbe4))



---
<sub><sup>*Generated with [git-changelog](https://github.com/rafinskipg/git-changelog). If you have any problem or suggestion, create an issue.* :) **Thanks** </sub></sup>

test/index.js
import { expect } from 'chai'
import Vreme from '../src/index'

let stamp

const date = new Date('Sat, 01 Aug 2015 2:10:21')

describe('Human Readable Time format', () => {

  before(() => stamp = new Vreme({
    usePrototype: true
  }))

  describe('Month conversion', () => {

    it('should convert date to full month name', () => expect(stamp.format(date, 'January')).to.equal('August'))

    it('should convert date to three letters month name', () => expect(stamp.format(date, 'Jan')).to.equal('Aug'))

    it('should convert date to all caps three letters month name', () => expect(stamp.format(date, 'JAN')).to.equal('AUG'))

    it('should convert date to lower case three letters month name', () => expect(stamp.format(date, 'jan')).to.equal('aug'))

  })

  describe('Day conversion', () => {

    it('should convert date to full day name', () => expect(stamp.format(date, 'Monday')).to.equal('Saturday'))

    it('should convert date to three letters day name', () => expect(stamp.format(date, 'Mon')).to.equal('Sat'))

    it('should convert date to two letters day name', () => expect(stamp.format(date, 'Mo')).to.equal('Sa'))

    it('should convert date to all caps two letters day name', () => expect(stamp.format(date, 'MO')).to.equal('SA'))

    it('should convert date to lowercase day name', () => expect(stamp.format(date, 'monday')).to.equal('saturday'))

    it('should convert date to ordinal day number (st)', () => expect(stamp.format(date, '1st')).to.equal('1st'))

    it('should convert date to ordinal day number (nd)', () => expect(stamp.format(new Date('Sun, 02 Aug 2015 2:10:21'), '1st')).to.equal('2nd'))

    it('should convert date to ordinal day number (rd)', () => expect(stamp.format(new Date('Mon, 03 Aug 2015 2:10:21'), '1st')).to.equal('3rd'))

    it('should convert date to ordinal day number (th)', () => expect(stamp.format(new Date('Tue, 04 Aug 2015 2:10:21'), '1st')).to.equal('4th'))

    it('should convert date to two numbers day format', () => expect(stamp.format(date, '24')).to.equal('01'))

  })

  describe('Year conversion', () => {

    it('should show 4 letter year', () => expect(stamp.format(date, '1999')).to.equal('2015'))

    it('should show 2 letter year for values between 32 and 99', () => expect(stamp.format(date, '35')).to.equal('15'))

  })

  describe('Time conversion', () => {

    it('should parse time in H:MM format', () => expect(stamp.format(date, '5:00')).to.equal('2:10'))

    it('should parse time in HH:MM format', () => expect(stamp.format(date, '17:10')).to.equal('02:10'))

    it('should parse time in H:MM:SS format', () => expect(stamp.format(date, '5:10:30')).to.equal('2:10:21'))

    it('should parse time in HH:MM:SS format', () => expect(stamp.format(date, '15:10:30')).to.equal('02:10:21'))

    it('should parse time in H:MM AM/PM format', () => expect(stamp.format(date, '5:10 pm')).to.equal('2:10 am'))

    it('should parse time in HH:MM AM/PM format for hours > 12', () => expect(stamp.format(new Date('Sat, 01 Aug 2015 14:10:21'), '11:10 pm')).to.equal('02:10 pm'))

    it('should parse time in HH:MMam/pm format', () => expect(stamp.format(date, '03:20pm')).to.equal('02:10am'))

  })

  describe('Combined conversions', () => {

    it('should format date in Monthname D, YYYY format', () => expect(stamp.format(date, 'March 25, 1999')).to.equal('August 01, 2015'))

    it('should format date in Monthname Dth format', () => expect(stamp.format(date, 'March 25th')).to.equal('August 1st'))

    it('should format date in YYYY/MM/DD format', () => expect(stamp.format(date, '2014/04/25')).to.equal('2015/08/01'))

    it('should format date in YY/MM/DD format', () => expect(stamp.format(date, '42/04/25')).to.equal('15/08/01'))

    it('should format date in MM/DD/YY format', () => expect(stamp.format(date, '10/10/10')).to.equal('08/01/15'))

    it('should format date in DD.MM.YYYY format', () => expect(stamp.format(date, '21.04.2015')).to.equal('01.08.2015'))

    it('should format date in YYYY/MM/DD HH:MM:SS format', () => expect(stamp.format(date, '21.04.2015 14:04:31')).to.equal('01.08.2015 02:10:21'))

    it('should format date in Monthname D, YYYY H:MM am/pm format', () => expect(stamp.format(date, 'March 25, 1999 2:04 PM')).to.equal('August 01, 2015 2:10 AM'))

    it('should keep the extraneous text', () => expect(stamp.format(date, 'Date: March, 25th')).to.equal('Date: August, 1st'))

    it('should keep the extraneous text after time', () => expect(stamp.format(date, 'Date: March, 25th 1:10pm lorem ipsum')).to.equal('Date: August, 1st 2:10am lorem ipsum'))

  });

  describe('Using prototype', () => it('should format date in YYYY/MM/DD format', () => expect(date.formatLike('2014/04/25')).to.equal('2015/08/01')))

});


src/index.js
export default class Vreme {

  constructor(options) {

    // Default options, should merge them with passed options in future
    this.options = {
      monthNames:   ['january', 'february', 'march', 'april', 'may', 'june', 'july', 'august', 'september', 'october', 'november', 'december'],
      dayNames:     ['sunday', 'monday', 'tuesday', 'wednesday', 'thursday', 'friday', 'saturday']
    }

    // Use Date prototype,
    // ie. `let date = new Date(); date.format_like('2015/11/22');`
    if (options && options.usePrototype) {
      // Arrow functions don't work here because we need `this` from
      // both class and function
      let self = this
      Date.prototype.formatLike = function(formatString) {
        return self.format(this, formatString)
      }
    }

    // Define all regular expressions
    this.regex = {
      MONTHNAMES_REGEXP:      new RegExp(`^(${this.options.monthNames.join('|')})$`, 'i'),
      MONTHNAMES_ABBR_REGEXP: new RegExp(`^(${this.options.monthNames.map(month => month.substr(0, 3)).join('|')})$`, 'i'),
      DAYNAMES_REGEXP:        new RegExp(`^(${this.options.dayNames.join('|')})$`, 'i'),
      DAYNAMES_ABBR_REGEXP:   new RegExp(`^(${this.options.dayNames.map(day => day.substr(0, 3)).join('|')})$`, 'i'),
      DAYNAMES_SHORT_REGEXP:  new RegExp(`^(${this.options.dayNames.map(day => day.substr(0, 2)).join('|')})$`, 'i'),
      ONE_DIGIT_REGEXP:       /^\d{1}$/,
      TWO_DIGIT_REGEXP:       /^\d{2}$/,
      FOUR_DIGIT_REGEXP:      /^\d{4}$/,
      ORDINAL_DAY_REGEXP:     /^(\d{1,2})(st|nd|rd|th)$/,
      TIME_REGEXP:            /(\d{1,2})(:)(\d{2})(\s*)(:)?(\d{2})?(\s*)?([ap]m)?/i
    }

    // Call reset method first, because we want a clean start
    this._reset()
  }

  // Main function
  format(date, formatString) {

    // Call reset method first, because we want a clean start
    this._reset()

    // Split format string by time
    let time   = formatString.split(this.regex.TIME_REGEXP)
    // Than get first element from array, because it should contain date
    let before = time.splice(0, 1)[0]
    // And the last one because it should stay the same
    let after  = time.pop()
    // When there's no time, last element is empty, so we need to check that too
    after = after || ''

    // Split date and format each part
    let formatedDate = before.split(/\b/).map(partial => this.formatDate(date, partial))
    let formatedTime = []

    if (time.length)
      formatedTime = time.map((partial, index) => this.formatTime(date, partial, index, time));

    // Merge and return the result
    // Time should be parsed too, but at the moment we are just returning it
    return formatedDate.join('') + formatedTime.join('') + after

  }

  // Handle ordinal suffixes
  // This could be a problem for other languages
  //
  // Args:
  //
  // - `day` (integer) - Day of the month
  //
  // Return:
  //
  // String representation of day with ordinal sufix
  ordinalSuffix(day) {

    let j = day % 10,
        k = day % 100

    if (j == 1 && k != 11)
      return day + 'st'

    if (j == 2 && k != 12)
      return day + 'nd'

    if (j == 3 && k != 13)
      return day + 'rd'

    return day + 'th'

  }

  // Format date partial
  //
  // Args:
  //
  // - `date` (Date) - Valid JavaScript date object
  // - `format` (String) - Part of a format string, use `format` for full string
  //
  // Return:
  //
  // Formated date partial (String)
  formatDate(date, format) {

    // Best match function
    let bestMatch = (number, date, twoDigits) => {
      // There is a better way to do this, but this should work now

      // If it's between 13 and 31 and day is not yet set
      if (number > 12 && number < 32 && !this.matches.day) {
        this.matches.day = true
        return ('0' + date.getDate()).slice(-2)
      }

      // If it's > 31 and year is not yet set
      if ((number > 12 && !this.matches.year) || number > 31) {
        this.matches.year = true
        return date.getFullYear() + ''
      }

      // Then first check if month is set
      if (!this.matches.month) {
        this.matches.month = true
        let month = date.getMonth() + 1
        if (twoDigits)
          return ('0' + month).slice(-2)
        return month + ''
      }

      // Then day
      if (!this.matches.day) {
        this.matches.day = true
        if (twoDigits)
          return ('0' + date.getDate()).slice(-2)
        return date.getDate() + ''
      }

      // And finally year
      if (!this.matches.year) {
        this.matches.year = true
        if (twoDigits)
          return ('0' + date.getFullYear()).slice(-2)
        return date.getFullYear() + ''
      }

      // Otherwise reset object
      this.reset()

      // Then call the function again
      return bestMatch(number, date)

    }

    // Check if format string is full month name
    if (format.match(this.regex.MONTHNAMES_REGEXP)) {
      this.matches.month = true
      return this._correctCase(format, this.options.monthNames[date.getMonth()])
    }

    // Check if format string is short month name
    if (format.match(this.regex.MONTHNAMES_ABBR_REGEXP)) {
      this.matches.month = true
      return this._correctCase(format, this.options.monthNames[date.getMonth()].substr(0, 3))
    }

    // Check if format string is full day name
    if (format.match(this.regex.DAYNAMES_REGEXP)) {
      this.matches.day = true
      return this._correctCase(format, this.options.dayNames[date.getDay()])
    }

    // Check if format string is 3 letter day name
    if (format.match(this.regex.DAYNAMES_ABBR_REGEXP)) {
      this.matches.day = true
      return this._correctCase(format, this.options.dayNames[date.getDay()].substr(0, 3))
    }

    // Check if format string is 2 letter day name
    if (format.match(this.regex.DAYNAMES_SHORT_REGEXP)) {
      this.matches.day = true
      return this._correctCase(format, this.options.dayNames[date.getDay()].substr(0, 2))
    }

    // Check if format string is year (4 digits)
    if (format.match(this.regex.FOUR_DIGIT_REGEXP)) {
      this.matches.year = true
      return date.getFullYear() + ''
    }

    // Check if format string is day with ordinal sufix
    if (format.match(this.regex.ORDINAL_DAY_REGEXP)) {
      this.matches.day = true
      return this.ordinalSuffix(date.getDate())
    }

    // This is a bit more complicated
    // If format string is two digits, then
    if (format.match(this.regex.TWO_DIGIT_REGEXP)) {
      // Parse it as an integer
      let number = parseInt(format, 10)

      // Check if it's obvious year
      if (number > 31 && number < 100) {
        this.matches.year = true
        return (date.getFullYear() + '').substring(2, 4)
      }

      // Otherwise try to find the best match (it could be year, month or day)
      return bestMatch(number, date, true)
    }

    // This is a bit more complicated too
    // If format string is just one digits, then
    if (format.match(this.regex.ONE_DIGIT_REGEXP)) {
      // Parse it as an integer
      let number = parseInt(format, 10)

      // And try to find the best match (it could be month, day or maybe year,
      // but that's a bit unlikelly)
      return bestMatch(number, date)
    }

    // Otherwise just return format, because it is probably just a text
    return format

  }

  formatTime(dateTime, format, index, fullTime) {

    let getAmPm = (format, hours) => {
      let ampm = hours < 12 ? 'am' : 'pm'
      if (this._isAllCaps(format))
        return ampm.toUpperCase()
      return ampm
    }

    if (index === 0 && format.match(this.regex.ONE_DIGIT_REGEXP))
      return dateTime.getHours()

    if (index === 0 && format.match(this.regex.TWO_DIGIT_REGEXP))
      if (fullTime[7] && dateTime.getHours() > 12) {
        return ('0' + (dateTime.getHours() - 12)).slice(-2)
      } else {
        return ('0' + dateTime.getHours()).slice(-2)
      }

    if (index === 2 && format.match(this.regex.TWO_DIGIT_REGEXP))
      return ('0' + dateTime.getMinutes()).slice(-2)

    if (format && index === 5 && format.match(this.regex.TWO_DIGIT_REGEXP))
      return ('0' + dateTime.getSeconds()).slice(-2)

    if (format && index === 7)
      return (dateTime.getHours() < 12) ? this._correctCase(format, 'am') : this._correctCase(format, 'pm')

    return format

  }

  _reset() {
    // Reset an object we are using for a best match
    this.matches = {
      month: false,
      day:   false,
      year:  false
    }
  }

  _isCapital(str) {
    return str.charAt(0) === str.charAt(0).toUpperCase()
  }

  _isLowercase(str) {
    return str === str.toLowerCase()
  }

  _isAllCaps(str) {
    return str === str.toUpperCase()
  }

  _toCapital(str) {
    return str.charAt(0).toUpperCase() + str.slice(1, str.length)
  }

  _correctCase(format, str) {
    if (this._isLowercase(format))
      return str.toLowerCase()
    if (this._isAllCaps(format))
      return str.toUpperCase()
    if (this._isCapital(format))
      return this._toCapital(str)
    return str
  }

}


vreme.js
'use strict';

(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports);
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports);
    global.index = mod.exports;
  }
})(this, function (exports) {
  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

  var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

  var Vreme = (function () {
    function Vreme(options) {
      var _this = this;

      _classCallCheck(this, Vreme);

      this.options = {
        monthNames: ['january', 'february', 'march', 'april', 'may', 'june', 'july', 'august', 'september', 'october', 'november', 'december'],
        dayNames: ['sunday', 'monday', 'tuesday', 'wednesday', 'thursday', 'friday', 'saturday']
      };

      if (options && options.usePrototype) {
        (function () {
          var self = _this;

          Date.prototype.formatLike = function (formatString) {
            return self.format(this, formatString);
          };
        })();
      }

      this.regex = {
        MONTHNAMES_REGEXP: new RegExp('^(' + this.options.monthNames.join('|') + ')$', 'i'),
        MONTHNAMES_ABBR_REGEXP: new RegExp('^(' + this.options.monthNames.map(function (month) {
          return month.substr(0, 3);
        }).join('|') + ')$', 'i'),
        DAYNAMES_REGEXP: new RegExp('^(' + this.options.dayNames.join('|') + ')$', 'i'),
        DAYNAMES_ABBR_REGEXP: new RegExp('^(' + this.options.dayNames.map(function (day) {
          return day.substr(0, 3);
        }).join('|') + ')$', 'i'),
        DAYNAMES_SHORT_REGEXP: new RegExp('^(' + this.options.dayNames.map(function (day) {
          return day.substr(0, 2);
        }).join('|') + ')$', 'i'),
        ONE_DIGIT_REGEXP: /^\d{1}$/,
        TWO_DIGIT_REGEXP: /^\d{2}$/,
        FOUR_DIGIT_REGEXP: /^\d{4}$/,
        ORDINAL_DAY_REGEXP: /^(\d{1,2})(st|nd|rd|th)$/,
        TIME_REGEXP: /(\d{1,2})(:)(\d{2})(\s*)(:)?(\d{2})?(\s*)?([ap]m)?/i
      };

      this._reset();
    }

    _createClass(Vreme, [{
      key: 'format',
      value: function format(date, formatString) {
        var _this2 = this;

        this._reset();

        var time = formatString.split(this.regex.TIME_REGEXP);
        var before = time.splice(0, 1)[0];
        var after = time.pop();
        after = after || '';
        var formatedDate = before.split(/\b/).map(function (partial) {
          return _this2.formatDate(date, partial);
        });
        var formatedTime = [];
        if (time.length) formatedTime = time.map(function (partial, index) {
          return _this2.formatTime(date, partial, index, time);
        });
        return formatedDate.join('') + formatedTime.join('') + after;
      }
    }, {
      key: 'ordinalSuffix',
      value: function ordinalSuffix(day) {
        var j = day % 10,
            k = day % 100;
        if (j == 1 && k != 11) return day + 'st';
        if (j == 2 && k != 12) return day + 'nd';
        if (j == 3 && k != 13) return day + 'rd';
        return day + 'th';
      }
    }, {
      key: 'formatDate',
      value: function formatDate(date, format) {
        var _this3 = this;

        var bestMatch = function bestMatch(number, date, twoDigits) {
          if (number > 12 && number < 32 && !_this3.matches.day) {
            _this3.matches.day = true;
            return ('0' + date.getDate()).slice(-2);
          }

          if (number > 12 && !_this3.matches.year || number > 31) {
            _this3.matches.year = true;
            return date.getFullYear() + '';
          }

          if (!_this3.matches.month) {
            _this3.matches.month = true;
            var month = date.getMonth() + 1;
            if (twoDigits) return ('0' + month).slice(-2);
            return month + '';
          }

          if (!_this3.matches.day) {
            _this3.matches.day = true;
            if (twoDigits) return ('0' + date.getDate()).slice(-2);
            return date.getDate() + '';
          }

          if (!_this3.matches.year) {
            _this3.matches.year = true;
            if (twoDigits) return ('0' + date.getFullYear()).slice(-2);
            return date.getFullYear() + '';
          }

          _this3.reset();

          return bestMatch(number, date);
        };

        if (format.match(this.regex.MONTHNAMES_REGEXP)) {
          this.matches.month = true;
          return this._correctCase(format, this.options.monthNames[date.getMonth()]);
        }

        if (format.match(this.regex.MONTHNAMES_ABBR_REGEXP)) {
          this.matches.month = true;
          return this._correctCase(format, this.options.monthNames[date.getMonth()].substr(0, 3));
        }

        if (format.match(this.regex.DAYNAMES_REGEXP)) {
          this.matches.day = true;
          return this._correctCase(format, this.options.dayNames[date.getDay()]);
        }

        if (format.match(this.regex.DAYNAMES_ABBR_REGEXP)) {
          this.matches.day = true;
          return this._correctCase(format, this.options.dayNames[date.getDay()].substr(0, 3));
        }

        if (format.match(this.regex.DAYNAMES_SHORT_REGEXP)) {
          this.matches.day = true;
          return this._correctCase(format, this.options.dayNames[date.getDay()].substr(0, 2));
        }

        if (format.match(this.regex.FOUR_DIGIT_REGEXP)) {
          this.matches.year = true;
          return date.getFullYear() + '';
        }

        if (format.match(this.regex.ORDINAL_DAY_REGEXP)) {
          this.matches.day = true;
          return this.ordinalSuffix(date.getDate());
        }

        if (format.match(this.regex.TWO_DIGIT_REGEXP)) {
          var number = parseInt(format, 10);

          if (number > 31 && number < 100) {
            this.matches.year = true;
            return (date.getFullYear() + '').substring(2, 4);
          }

          return bestMatch(number, date, true);
        }

        if (format.match(this.regex.ONE_DIGIT_REGEXP)) {
          var number = parseInt(format, 10);
          return bestMatch(number, date);
        }

        return format;
      }
    }, {
      key: 'formatTime',
      value: function formatTime(dateTime, format, index, fullTime) {
        var _this4 = this;

        var getAmPm = function getAmPm(format, hours) {
          var ampm = hours < 12 ? 'am' : 'pm';
          if (_this4._isAllCaps(format)) return ampm.toUpperCase();
          return ampm;
        };

        if (index === 0 && format.match(this.regex.ONE_DIGIT_REGEXP)) return dateTime.getHours();
        if (index === 0 && format.match(this.regex.TWO_DIGIT_REGEXP)) if (fullTime[7] && dateTime.getHours() > 12) {
          return ('0' + (dateTime.getHours() - 12)).slice(-2);
        } else {
          return ('0' + dateTime.getHours()).slice(-2);
        }
        if (index === 2 && format.match(this.regex.TWO_DIGIT_REGEXP)) return ('0' + dateTime.getMinutes()).slice(-2);
        if (format && index === 5 && format.match(this.regex.TWO_DIGIT_REGEXP)) return ('0' + dateTime.getSeconds()).slice(-2);
        if (format && index === 7) return dateTime.getHours() < 12 ? this._correctCase(format, 'am') : this._correctCase(format, 'pm');
        return format;
      }
    }, {
      key: '_reset',
      value: function _reset() {
        this.matches = {
          month: false,
          day: false,
          year: false
        };
      }
    }, {
      key: '_isCapital',
      value: function _isCapital(str) {
        return str.charAt(0) === str.charAt(0).toUpperCase();
      }
    }, {
      key: '_isLowercase',
      value: function _isLowercase(str) {
        return str === str.toLowerCase();
      }
    }, {
      key: '_isAllCaps',
      value: function _isAllCaps(str) {
        return str === str.toUpperCase();
      }
    }, {
      key: '_toCapital',
      value: function _toCapital(str) {
        return str.charAt(0).toUpperCase() + str.slice(1, str.length);
      }
    }, {
      key: '_correctCase',
      value: function _correctCase(format, str) {
        if (this._isLowercase(format)) return str.toLowerCase();
        if (this._isAllCaps(format)) return str.toUpperCase();
        if (this._isCapital(format)) return this._toCapital(str);
        return str;
      }
    }]);

    return Vreme;
  })();

  exports.default = Vreme;
});
